libraries/shared/storage/backfill/storage_value_loader.go

```diff
@@ -0,0 +1,111 @@
+	"github.com/makerdao/vulcanizedb/libraries/shared/storage/types"
+	"github.com/makerdao/vulcanizedb/libraries/shared/transformer"
+	"github.com/makerdao/vulcanizedb/pkg/core"
+	"github.com/makerdao/vulcanizedb/pkg/datastore"
+	"github.com/makerdao/vulcanizedb/pkg/datastore/postgres"
+	"github.com/makerdao/vulcanizedb/pkg/datastore/postgres/repositories"
+	"github.com/sirupsen/logrus"
+)
+
+func NewStorageValueLoader(bc core.BlockChain, db *postgres.DB, initializers []transformer.StorageTransformerInitializer, blockNumber int64) StorageValueLoader {
+	return StorageValueLoader{
+		bc:              bc,
+		db:              db,
+		HeaderRepo:      repositories.NewHeaderRepository(db),
+		StorageDiffRepo: storage2.NewDiffRepository(db),
+		initializers:    initializers,
+		blockNumber:     blockNumber,
+	}
+}
+
+type StorageValueLoader struct {
+	bc              core.BlockChain
+	db              *postgres.DB
+	HeaderRepo      datastore.HeaderRepository
+	StorageDiffRepo storage2.DiffRepository
+	initializers    []transformer.StorageTransformerInitializer
+	blockNumber     int64
+}
+
+func (r *StorageValueLoader) Run() error {
+	addressToKeys, getKeysErr := r.getStorageKeys()
+	if getKeysErr != nil {
+		return getKeysErr
+	}
+	logrus.Infof("Received storage keys for %v addresses.", len(addressToKeys))
+
+	header, getHeaderErr := r.HeaderRepo.GetHeader(r.blockNumber)
+	if getHeaderErr != nil {
+		return getHeaderErr
+	}
+
+	for address, keys := range addressToKeys {
+		persistStorageErr := r.getAndPersistStorageValues(address, keys, r.blockNumber, header.Hash)
+		if persistStorageErr != nil {
+			return persistStorageErr
+		}
+	}
+	logrus.Infof("Persisted storage values for %v addresses.", len(addressToKeys))
+
+	return nil
+}
+
+func (r *StorageValueLoader) getAndPersistStorageValues(address common.Address, keys []common.Hash, blockNumber int64, headerHash string) error {
+	blockNumberBigInt := big.NewInt(blockNumber)
+	for _, key := range keys {
+		value, getStorageErr := r.bc.GetStorageAt(address, key, blockNumberBigInt)
+		if getStorageErr != nil {
+			return getStorageErr
+		}
+		diff := types.RawDiff{
+			HashedAddress: crypto.Keccak256Hash(address[:]),
+			BlockHash:     common.HexToHash(headerHash),
+			BlockHeight:   int(blockNumber),
+			StorageKey:    key,
+			StorageValue:  common.BytesToHash(value),
+		}
+
+		diffId, createDiffErr := r.StorageDiffRepo.CreateStorageDiff(diff)
+		if createDiffErr != nil {
+			if createDiffErr == sql.ErrNoRows {
+				return nil
+			}
+			return createDiffErr
+		}
+
+		markFromBackfillErr := r.StorageDiffRepo.MarkFromBackfill(diffId)
+		if markFromBackfillErr != nil {
+			return markFromBackfillErr
+		}
+	}
+	return nil
+}
+
+func (r *StorageValueLoader) getStorageKeys() (map[common.Address][]common.Hash, error) {
+	addressToKeys := make(map[common.Address][]common.Hash)
+	for _, i := range r.initializers {
+		transformer := i(r.db)
+		keysLookup, ok := transformer.GetStorageKeysLookup().(storage.KeysLookup)
```

Comment:
does this cast potentially cause a nil pointer panic if `!ok`?

Reply:
I think that `ok` is a bool, so by capturing that in a var we'll able to handle it and return an error instead of panicking. Not sure how to test this out in the code itself, but tried it out in a contrived example here: https://play.golang.org/p/UUfNh5tnvLr